File: rungekutta.py
#runge-kutta fourth-order numerical integrater

import math

class rungekutta:
    def __init__(self, steps):
        self.steps = steps
    
    def rungekutta(self, func, initx, inity, evaluatedx):
        h = (evaluatedx - initx) / self.steps
        for _ in range(self.steps):
            k1 = h * func(initx, inity)
            k2 = h * func(initx + 0.5 * h, inity + 0.5 * k1)
            k3 = h * func(initx + 0.5 * h, inity + 0.5 * k2)
            k4 = h * func(initx + h, inity + k3)
            evaluatedy = inity + (1/6)*(k1 + 2*k2 + 2*k3 + k4)
            inity = evaluatedy
            initx += h
        return evaluatedy

    def numericalIntegrate(self, func, xi, xf):
        initx = 0
        inity = func(initx)
        return self.rungekutta(lambda x,y: func(x), initx, inity, xf) - self.rungekutta(lambda x,y: func(x), initx, inity, xi)

    def polynomialIntegrate(self, xi, xf, *args):
        polynomial = []
        for coefs in args:
            polynomial.append(coefs)
        def func(x):
            sum = 0
            for i in range(len(polynomial)):
                sum += polynomial[i] * x**(len(polynomial) - i - 1)
            return sum
        initx = 0
        inity = func(initx)
        return self.rungekutta(lambda x,y: func(x), initx, inity, xf) - self.rungekutta(lambda x,y: func(x), initx, inity, xi)
    
#tests
integrater = rungekutta(steps=100)
print(integrater.rungekutta(lambda x,y: x**2+y, 4, 6, 10)) #solution to dy/dx=x^2+y(x) with initial condition y(4)=6 at x=10
print(integrater.numericalIntegrate(lambda x: math.sin(x), 5, 10)) #definite integral of sin(x) from x=5 to x=10
print(integrater.polynomialIntegrate(5,10,2,4,3,6)) #definite integral of 2x^3+4x^2+3x+6 from x=5 to x=10

File: rungekutta.c
//runge-kutta fourth-order numerical integrater

#include <stdio.h>
#include <math.h>

double rungekutta(double (*f)(double, double), double* initx, double* inity, double evaluatedx, int steps){
    double h = (evaluatedx - *initx) / steps;
    double evaluatedy;
    for(int i = 0; i < steps; i++){
        double k1 = h * (*f)(*initx, *inity);
        double k2 = h * (*f)(*initx + 0.5 * h, *inity + 0.5 * k1);
        double k3 = h * (*f)(*initx + 0.5 * h, *inity + 0.5 * k2);
        double k4 = h * (*f)(*initx + h, *inity + k3);
        evaluatedy = *inity + (1/6.0)*(k1 + 2*k2 + 2*k3 + k4);
        *inity = evaluatedy;
        *initx += h;
    }
    return evaluatedy;
}

double numericalIntegrate(double (*f)(double, double), double xi, double xf, int steps){
    double initx = 0.0;
    double inity = (*f)(initx, 0);
    return rungekutta((*f), &initx, &inity, xf, steps) - rungekutta((*f), &initx, &inity, xi, steps);
}

double function(double x, double y){
    return pow(x,4);
}

double function2(double x, double y){
    return x; //do not use the y parameter here - the numerical integrater is for one variable only.
}


int main(){
    double initx = 0;
    double inity = 0;
    double evaluatedx = 5;
    int steps = 20;
    printf("The solution to dy/dx (%d steps) with y(%lf)=%lf is y(%lf)=%lf\n", steps, initx, inity, evaluatedx, rungekutta(function, &initx, &inity, evaluatedx, steps));
    printf("%lf\n", numericalIntegrate(function2, 7, 10, 10));
    return 0;
}
